DOC: interpolate: add tutorial on how to fill an image from its boundary

DWesl opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message.
However, please only include an issue number in the description, not the title,
and please ensure that any code names containing underscores are enclosed in backticks.

Depending on your changes, you can skip CI operations and save time and energy: 
https://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->
Related to #21910
Replaces #23285

#### What does this implement/fix?
<!--Please explain your changes.-->

This tutorial documents ways to estimate the interior of an image given only values on the boundary.  Much of that consists of reducing the input or requested points, to avoid NaNs or to reduce computational time.

The simplex tolerance section depends on #23737, and should probably be commented out until that is merged.

#### Additional information
<!--Any additional information you think is important.-->

The original query was about estimating the background illumination on a microscope slide.

[docs only]

DWesl committed:
DOC: Add a tutorial covering how to fill an image from its boundary.

The original query was about estimating the background illumination on a microscope slide.

ev-br commented on doc/source/tutorial/interpolate/2D_fill_from_border.rst:
```diff
@@ -0,0 +1,408 @@
+------------------
+
+With this simple boundary condition, we can find a few analytic
+solutions as a baseline.
```
I'm not entirely sure what is the problem these are solutions for. You mentioned estimating the background above, maybe expand it a little? This whole story started from a problem, maybe discuss it a bit? Even if you start from a real-world case and then quickly turn into a simplified artificial example.

ev-br commented on doc/source/tutorial/interpolate/2D_fill_from_border.rst:
```diff
@@ -0,0 +1,408 @@
+
+
+Statsmodels offers a regression interface that hides the linear
+algebra and offers additional diagnostics.
```
Again, would be nice to have a link to a specific statsmodels function(s), and some reasoning to which statsmodels functionality I as a reader should be looking at and why.

ev-br reviewed (commented):
Looks nice on a cursory glance!
Will do a more careful read-through later. Meanwhile, it would be nice to understand and see
- how does this relate to other tutorial pages, where should this be linked from
- (for a non-expert like me) what is the problem(s) this solves? Is there a way to have expand the motivation / what are the requirements / what are the constraints?
- is there a way to relate the material in this tutorial to a real-world problem (e.g. the one that started https://github.com/scipy/scipy/issues/21910, was it something about microscopy images?

DWesl committed:
DOC: Extend boundary documentation.

DWesl commented:
> We can adopt a Diataxis if that helps (not sure what that means but no matter, the structure of scipy.interpolate tutorials is not set in stone).

The relevant bit is division into tutorial, how-to, explanation, and reference, based on whether they're designed to help acquire or apply knowledge and another axis I can't summarize as easily, but discussion of whether SciPy should adopt [NEP 44](https://www.numpy.org/neps/nep-0044-restructuring-numpy-docs) should probably be left for the mailing list.

> > I suspect I need to think about that more clearly. The contrived example is the kind I'm most familiar with, but anyone with _that kind of problem_ will likely come across it in the context of solving a Laplace equation and will go for Laplace solvers before SciPy.
> 
> (emphasis mine) My actual question is what is that kind of problem? Just reading what's written leaves me (a non-expert!) very confused:  if we're talking about the boundary value problem for the Laplace equation, then it's a problem with a theoretically unique solution (provided some conditions are met etc). 

That's the bit I need to think about.  My first reaction on seeing "I have values near the edges, and want values near the center", especially with the rectangular domain, is to treat is at a Laplace problem.  The theoretical guarantee that a solution exists and is unique is great ... if the Laplace equation describes what they want.  Laplace solutions tend to be *very* smooth, which might be what any given user wants, or might not be.

> If it's _estimating the background_, then is there a specific model for the background, or what are the constraints for the estimation? (Or does the Laplace BVP appear as an example where we just know the correct solution).

For the general case, I can't tell.  Part of what I was trying to do was to show that you need additional constraints to get a unique solution, and show some of those constraints

If the Laplace equation is a poor match to their problem, my next conceptualization of the problem is "I'm trying to predict a value at these points based on these other points", which is a regression problem, and polynomial regression is reasonably straightforward conceptually (I can also write those examples using `optimize.curve_fit` instead of rolling my own linear regression.

For the microscope illumination problem, I could find an equation for diffuse reflection of a point source from a plane toward a point observer and try to fit the five parameters, but that feels like a better fit for a microscopy package than for SciPy.

The lack of an actual problem statement is part of why I made #23285 first.  Elaborating on why `interpolate.griddata` is a weird choice for whatever problem results in information about the edges but not the middle and a need for the middle is why I made this PR.

YuzeHao2023 reviewed (approved):
lgtm

ev-br reviewed (commented):
> discussion of whether SciPy should adopt NEP 44 should probably be left for the mailing list.

Agreed. A clearer split would be nice indeed, needs a champion :-). Whether to start from a ML or from reorganizing existing tutorial pages for one of submodules... Up to that champion, I suppose. 

> division into tutorial, how-to, explanation, and reference, based on whether they're designed to help acquire or apply knowledge

So what you're adding is a how-to, it seems. Which is great! So it's somewhat similar to the set of working examples in https://docs.scipy.org/doc/scipy/tutorial/interpolate/extrapolation_examples.html 
(written without relation to NEP 44 or anything diataxis though).

DWesl committed:
Merge branch 'main' of https://github.com/scipy/scipy into infill-tutorial

DWesl committed:
DOC: Add 2-D infill tutorial to interpolate index

Forgot this step earlier.

[docs only]

DWesl committed:
Merge branch 'scipy:main' into infill-tutorial

DWesl committed:
Merge branch 'infill-tutorial' of DWesl/scipy into infill-tutorial

DWesl committed:
DOC,BUG: Reduce sig figs printed for error.

Many are near machine epsilon, so digits might differ between machines.
This caused a doc build failure.

[docs only]

DWesl committed:
DOC,FIX: Update RBF mean errors in 2D interpolation tutorial

The linear error was larger, the quintic error was smaller.

The most recent evaluation was with Python 3.9 and SciPy<1.14

[docs only]

DWesl committed:
DOC: Extend the introduction to the border-filling tutorial.

[docs only]

DWesl committed:
Merge branch 'infill-tutorial' of DWesl/scipy into infill-tutorial

DWesl committed:
DOC: Switch example output to array-likes.

A bit more work for humans to understand,
but much easier for the automated checkers.

[docs only]

DWesl committed:
DOC,FIX: Update outputs for Python 3.12.

Mean values for the regressions have changed slightly since 3.9.

Suppresses some of the axes printed.

DWesl committed:
DOC,STY: Wrap lines in examples to avoid scrollbars.

I dislike scrollbars and have a moderately wide screen.
This should make the examples easier to read for others.

DWesl committed:
DOC,FIX: Fix syntax for links to other packages.

I used Markdown syntax instead of ReStructured Text.

DWesl committed:
DOC: Adjust plots so subplots are larger.

Set constrained_layout=True to ensure labels stay in visible area.
[docs only]

DWesl committed:
STY: Fix some formatting and typos, add versionadded for simplex_tolerance.

DWesl committed:
Merge branch 'main' into infill-tutorial

DWesl committed:
Merge branch 'main' of scipy/scipy into infill-tutorial

DWesl committed:
DOC,TST: Drop quintic kernel from RBF examples.

[docs only]
It seems to be somewhat ill-conditioned,
enough for roundoff error got get to 1e-4 or so.
The problem setup likely isn't helping.

DWesl committed:
DOC: Specify sizes for all figures in 2D interpolation tutorial.

[docs only]

ev-br reviewed (approved):
Sorry for the delay and thank you for the write-up!
This now reads great, and shows multiple tricks and tips. I wish this existed when I started looking at 2D interpolators in SciPy.

The only thing which stands between me and the green button is 20 commits, which is a bit much. Is there some history you'd want to preserve @DWesl? If so could you please squash it a bit. Otherwise, I'll squash-merge.

DWesl committed:
DOC: Add closing notes and some section headers.

Also remove references to quintic RBF kernel,
since that got removed for numerical reasons.

[docs only]

ev-br commented:
Squash-merged, thank you @DWesl , much appreciated!
